versioned/spi/src/main/java/org/projectnessie/versioned/NamedReference.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned;
+
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface NamedReference<METADATA> {
+  NamedRef getNamedRef();
+
+  Hash getHash();
+
+  @Nullable
+  Hash getCommonAncestor();
+
+  NamedReference<METADATA> withCommonAncestor(@Nullable Hash value);
+
+  @Nullable
+  CommitsAheadBehind getAheadBehind();
+
+  NamedReference<METADATA> withAheadBehind(@Nullable CommitsAheadBehind value);
+
+  @Nullable
+  METADATA getHeadCommitMeta();
+
+  NamedReference<METADATA> withHeadCommitMeta(@Nullable METADATA value);
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  default <UPDATED_METADATA> NamedReference<UPDATED_METADATA> withUpdatedCommitMeta(
+      UPDATED_METADATA commitMeta) {
+    NamedReference updated = this;
+    return updated.withHeadCommitMeta(commitMeta);
```

Comment:
Technically, we're breaking the API contract of `this` here. If its accessors were to be called via an old reference with the old type parameter, it would lead to a class cast exception.

Could we create a separate `NamedReference` object here? For example:
```
default <UPDATED_METADATA> NamedReference<UPDATED_METADATA> reinterpretCommitMeta(Function<METADATA, UPDATED_METADATA> parser) {
   return builder().from(this)...
}
```
WDYT?

Reply:
Why should it cause a CCE?
It's setting the commit-meta to an instance of `UPDATED_METADATA`, so the old one's is gone.

(I've only added this function so that call sites do not have to deal with raw types.)

Reply:
I mean there's no problem in existing code that uses `NamedReference`. My comment was "in principle" :) ... something like this:

```
NamedReference<ByteString> ref0 = ...
NamedReference<CommitMeta> ref1 = ref0.withUpdatedCommitMeta(...);
ByteString rawBytes = ref0.getCommitMeta(); // compiler is happy here, but CCE in runtime, AFAIK
```

Reply:
No problem. That `with*` creates a copy

Reply:
Ah, yes, I missed the copy in generated `with` methods :facepalm: